src/System.IO.FileSystem/src/System/IO/Win32FileSystemEnumerable.cs

```diff
@@ -513,182 +535,155 @@ private static String GetFullSearchString(String fullPath, String searchPattern)
+        internal static SearchResultHandler<DirectoryInfo> DirectoryInfo
         {
-            String name = result.FullPath;
-            IFileSystemObject fileSystemObject = new Win32FileSystem.Win32FileSystemObject(name, result.FindData, asDirectory: true);
-            DirectoryInfo di = new DirectoryInfo(name, fileSystemObject);
-            return di;
+            get { return s_directoryInfo ?? (s_directoryInfo = new DirectoryInfoResultHandler()); }
         }
-    }
 
-    internal class FileSystemInfoResultHandler : SearchResultHandler<FileSystemInfo>
-    {
-        [System.Security.SecurityCritical]
-        internal override bool IsResultIncluded(SearchResult result)
+        internal static SearchResultHandler<FileSystemInfo> FileSystemInfo
         {
-            bool includeFile = Win32FileSystemEnumerableHelpers.IsFile(result.FindData);
-            bool includeDir = Win32FileSystemEnumerableHelpers.IsDir(result.FindData);
-            Debug.Assert(!(includeFile && includeDir), result.FindData.cFileName + ": current item can't be both file and dir!");
-
-            return (includeDir || includeFile);
+            get { return s_fileSystemInfo ?? (s_fileSystemInfo = new FileSystemInfoResultHandler()); }
         }
 
-        [System.Security.SecurityCritical]
-        internal override FileSystemInfo CreateObject(SearchResult result)
+        private sealed class StringResultHandler : SearchResultHandler<string>
         {
-            bool isFile = Win32FileSystemEnumerableHelpers.IsFile(result.FindData);
-            bool isDir = Win32FileSystemEnumerableHelpers.IsDir(result.FindData);
-            String name = result.FullPath;
+            private readonly bool _includeFiles;
+            private readonly bool _includeDirs;
 
-            if (isDir)
+            internal StringResultHandler(bool includeFiles, bool includeDirs)
             {
-                IFileSystemObject fileSystemObject = new Win32FileSystem.Win32FileSystemObject(name, result.FindData, asDirectory: true);
-                DirectoryInfo di = new DirectoryInfo(name, fileSystemObject);
-                return di;
+                _includeFiles = includeFiles;
+                _includeDirs = includeDirs;
             }
-            else
+
+            [SecurityCritical]
+            internal override bool IsResultIncluded(string fullPath, string userPath, ref Interop.mincore.WIN32_FIND_DATA findData, out string result)
             {
-                Debug.Assert(isFile);
-                IFileSystemObject fileSystemObject = new Win32FileSystem.Win32FileSystemObject(name, result.FindData, asDirectory: false);
-                FileInfo fi = new FileInfo(name, fileSystemObject);
-                return fi;
+                if (_includeFiles && Win32FileSystemEnumerableHelpers.IsFile(ref findData) ||
+                    _includeDirs && Win32FileSystemEnumerableHelpers.IsDir(ref findData))
```

Comment:
I know the parens aren't strictly needed here due to && having higher precedence than ||, but I still find it much easier to understand if there are parents around each grouping:
```C#
if ((_includeFiles && ...) ||
    (_includeDirs && ...))
{
    ...
}
```